src/systems/world_exporter/WorldExporter.cc

```diff
@@ -0,0 +1,331 @@
+      if (material != nullptr)
+      {
+        mat->SetDiffuse(material->Data().Diffuse());
+        mat->SetAmbient(material->Data().Ambient());
+        mat->SetEmissive(material->Data().Emissive());
+        mat->SetSpecular(material->Data().Specular());
+      }
+      mat->SetTransparency(_transparency->Data());
+
+      const ignition::common::Mesh * mesh;
+      std::weak_ptr<ignition::common::SubMesh> subm;
+      math::Vector3d scale;
+      math::Matrix4d matrix(localPose);
+      ignition::common::MeshManager *meshManager =
+          ignition::common::MeshManager::Instance();
+
+      if (_geom->Data().Type() == sdf::GeometryType::BOX)
+      {
+        igndbg << "It is a Box!"  << std::endl;
+        if (meshManager->HasMesh("unit_box"))
+        {
+          igndbg << "Box Found..."  << std::endl;
+          mesh = meshManager->MeshByName("unit_box");
+          scale = _geom->Data().BoxShape()->Size();
+          int i = worldMesh.AddMaterial(mat);
+
+          subm = worldMesh.AddSubMesh(
+            *mesh->SubMeshByIndex(0).lock().get());
+          subm.lock()->SetMaterialIndex(i);
+          subm.lock()->Scale(scale);
+          subMeshMatrix.push_back(matrix);
+        }
+        else
+        {
+          igndbg << "Box NOT Found"  << std::endl;
+        }
+      }
+      else if (_geom->Data().Type() == sdf::GeometryType::CYLINDER)
+      {
+        if (meshManager->HasMesh("unit_cylinder"))
+        {
+          igndbg << "Cylinder Found"  << std::endl;
+          mesh = meshManager->MeshByName("unit_cylinder");
+          scale.X() = _geom->Data().CylinderShape()->Radius() * 2;
+          scale.Y() = scale.X();
+          scale.Z() = _geom->Data().CylinderShape()->Length();
+
+          igndbg << "Pose : x: " << localPose << std::endl;
+
+          int i = worldMesh.AddMaterial(mat);
+
+          subm = worldMesh.AddSubMesh(
+            *mesh->SubMeshByIndex(0).lock().get());
+          subm.lock()->SetMaterialIndex(i);
+          subm.lock()->Scale(scale);
+          subMeshMatrix.push_back(matrix);
+        }
+        else
+        {
+          igndbg << "Cylinder NOT Found"  << std::endl;
+        }
+      }
+      else if (_geom->Data().Type() == sdf::GeometryType::PLANE)
+      {
+        igndbg << "It is a Plane!"  << std::endl;
+        if (meshManager->HasMesh("unit_plane"))
+        {
+          igndbg << "Plane Found"  << std::endl;
+          // Create a rotation for the plane mesh to account
+          // for the normal vector.
+          mesh = meshManager->MeshByName("unit_plane");
+
+          scale.X() = _geom->Data().PlaneShape()->Size().X();
+          scale.Y() = _geom->Data().PlaneShape()->Size().Y();
+
+          // // The rotation is the angle between the +z(0,0,1) vector and the
+          // // normal, which are both expressed in the local (Visual) frame.
+          math::Vector3d normal = _geom->Data().PlaneShape()->Normal();
+          localPose.Rot().From2Axes(
```

Comment:
My idea is that localPose (now worldPose) already has in it the rotation and translation of the view, so I'll add on top of it the rotation of the normal and that is what gets exported to the matrix.

Reply:
> I'll add on top of it the rotation of the normal and that is what gets exported to the matrix

Yes, I think that's what we want, but I don't think that's what's happening. The current code is
```c++
worldPose.Rot().From2Axes(
    math::Vector3d::UnitZ, normal.Normalized());

matrix = math::Matrix4d(worldPose);
```
I believe this will replace the rotation in `worldPose` with the rotation calculated from the normal. I don't think it applies it on top of the current rotation in `worldPose`. Here's my suggestion

```c++
math::Quaterniond normalRot;
normalRot.From2Axes(math::Vector3d::UnitZ, normal.Normalized());
worldPose.Rot() = worldPose.Rot() * normalRot;
```

Reply:
Updated in 3206ad83e13221d83349cf5d44655f382f45a8e6